keychain.h

```diff
@@ -0,0 +1,70 @@
+#include <assert.h>
+
+#include <optional>
+#include <string>
+#include <variant>
+
+namespace keychain {
+
+class Keychain {
+   public:
+    struct Error {
+        std::string message;
+    };
+    using Success = std::monostate;
+    using Password = std::string;
+    using Result = std::variant<Success, Password, Error>;
+
+    static Result getPassword(const std::string &service,
+                              const std::string &user);
+    static Result setPassword(const std::string &service,
+                              const std::string &user,
+                              const std::string &password);
+    static Result deletePassword(const std::string &service,
+                                 const std::string &user);
+
+   public:
+    /*! \brief Construct a Keychain specific to a service/user combination
+     *
+     * \param service Identifier for the service whose password should be stored
+     *                or updated. Note that a platform specific service name
+     *                will be used based on the provided identifier.
+     *                \sa getServiceName
+     * \param user
+     */
+    Keychain(const std::string &service, const std::string &user);
+
+    std::optional<std::string> getPassword();
+    bool setPassword(const std::string &password);
+    bool deletePassword();
+
+    //! Return the error resulting from the most recent call to getPassword,
+    //! addPassword, or deletePassword, if any
+    std::optional<std::string> lastError();
+
+   private:
+    //! \brief Create a platform specific service identifier
+    std::string makeServiceName(const std::string &service);
+
+    void clearError();
+
+    template <typename F>
+    bool updatePassword(F &&updateFun) {
```

Comment:
Pass by value should be fair enough here. By rvalue reference would allow you to `std::forward` which you don’t do here...

Reply:
Oops. I think i'll `forward`.

Reply:
But this is actually not an rvalue reference but a forwarding reference (formerly known as universal reference, I believe). If you make it a normal value, a call like that:

```C++
updatePassword([] { /* ... */ });
```

will actually just instantiate `updateFun` in the function call once during the function call.

Reply:
As discussed in person, the forwarding ref seems to be the be the best option here.